lib/complex.ex

```diff
@@ -0,0 +1,29 @@
+defmodule Complex do
+  @moduledoc """
+  A simple attempt at complex numbers
+  """
+  defstruct re: 0, im: 0
+
+  def new({re, im}), do: new(re, im)
+  def new(re, im) when is_number(re) and is_number(im) do
+    %__MODULE__{re: re, im: im}
+  end
+
+  def add(a = %__MODULE__{}, b = %__MODULE__{}) do
```

Comment:
As an alternative to pattern matching on structs, you could look into typespecs in combination with the tool Dialyzer.

http://elixir-lang.org/getting-started/typespecs-and-behaviours.html

That said, if you don't pattern match, you'll allow any map that has the keys `re` and `im`, which may or may not be what you intend. I guess it boils down to how much or little duck-typing you're after. :-)

For this particular case, keeping the pattern matching seems fine to me. Adding typespecs will be icing on the cake and a fun exercise.